### PR TITLE
Disable dynamo if a closure is present

### PR DESCRIPTION
**Pull request by mlazos**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #123469
* #123406
* __->__ #123476
* #123405
* #123498
* #123404
* #123324
* #123323
* #123322
* #123620
* #123619
* #123618
* #123552
* #123551
* #123497
* #123496

### CODE REVIEW

File: torch/optim/optimizer.py [hunk 1]
```diff
@@ -54,6 +54,15 @@ def __repr__(self) -> str:
 required = _RequiredParameter()
 
 
+def _disable_dynamo_if_closure(fn):
+    def inner(self, *args, **kwargs):
+        if (len(args) == 1 and args[0] is not None) or (len(kwargs) == 1 and kwargs.get("closure", None) is not None):
+            return torch._disable_dynamo(functools.partial(fn))(self, *args, **kwargs)
```
Comment (anijain2305) [inline torch/optim/optimizer.py:60]: How about simply

`return torch._disable_dynamo(fn)(self, *args, **kwargs)`

Comment (mlazos) [reply torch/optim/optimizer.py:60]: that will disable the step forever, won't it?

Comment (anijain2305) [reply torch/optim/optimizer.py:60]: I don't get the last comment. If you use two optimizer instances, both will have their own decorators.

Comment (mlazos) [reply torch/optim/optimizer.py:64]: won't step be disabled forever for that instance? I was under the impression that it mutated the function that is disabled (ie if recursive is False I see skip and stuff being set) if recursive=True does it not mutate the function instance?

Comment (mlazos) [reply torch/optim/optimizer.py:64]: yeah you right, just tried it, doesn't look like it mutates, I'll remove

**Comment by mlazos** [comment]
Figured out other ways around this issue, we trace it now. The frame was getting skipped, bypassing the optimizer handling

**Pull request closed without merging**